From 6ad58f2cb802d7596815ade9d6221cbac5ea34d9 Mon Sep 17 00:00:00 2001
From: Abhishek Chauhan <quic_abchauha@quicinc.com>
Date: Sun, 12 Oct 2025 20:34:37 -0700
Subject: [PATCH 11/16] PENDING: net: PCS: Check and recover pcs link after
 serdes reset

Currently we dont have reset domain for serdes, so there is no
way for kernel space to reset the serdes which is programmed in the
SCMI platform. With this flexibility to reset serdes from emac
driver added it is good to check for XPCS link status.

After serdes is reset from the emac block, according to programming
guidelines, it is required to check the XPCS link to ensure no
BLK_ERR, BLOCK_LOCK errors and RLU failures have occurred. Retry
if required by resetting the PCS block to ensure group alignment.

Change-Id: I359aece38ad1e094cc465d09cf477b721ae65d70
Signed-off-by: Abhishek Chauhan <quic_abchauha@quicinc.com>
(cherry picked from commit bcd1d610a4181b8e75364ad6f34f946a80157d3c)
---
 drivers/net/pcs/pcs-xpcs-qcom.c | 91 +++++++++++++++++++++++++++++++--
 drivers/net/pcs/pcs-xpcs-qcom.h |  6 ++-
 2 files changed, 93 insertions(+), 4 deletions(-)

diff --git a/drivers/net/pcs/pcs-xpcs-qcom.c b/drivers/net/pcs/pcs-xpcs-qcom.c
index 1fd0cacc5601..93a8535665fc 100644
--- a/drivers/net/pcs/pcs-xpcs-qcom.c
+++ b/drivers/net/pcs/pcs-xpcs-qcom.c
@@ -26,6 +26,9 @@
 #define XPCSERR(fmt, args...) \
 	pr_err(DRV_NAME " %s:%d " fmt, __func__, __LINE__, ## args)
 
+#define XPCSDBG(fmt, args...) \
+	pr_debug(DRV_NAME " %s:%d " fmt, __func__, __LINE__, ## args)
+
 static const int xpcs_usxgmii_features[] = {
 	ETHTOOL_LINK_MODE_Pause_BIT,
 	ETHTOOL_LINK_MODE_Asym_Pause_BIT,
@@ -445,6 +448,80 @@ static int qcom_xpcs_config(struct phylink_pcs *pcs, unsigned int mode, phy_inte
 	return qcom_xpcs_do_config(qxpcs, interface);
 }
 
+static int qcom_xpcs_get_link_status(struct dw_xpcs_qcom *qxpcs,
+				     struct phylink_link_state *state)
+{
+	unsigned int retries = 32;
+	unsigned int count = 0;
+	int ret = -EFAULT;
+
+	if (!qxpcs)
+		goto failure;
+
+/* Try to recover the link before posting link failure */
+recover:
+	count++;
+
+	if (count >= retries) {
+		XPCSERR("Link recovery failed\n");
+		goto failure;
+	}
+
+	/* Reset the link to make sure XPCS is aligned to Serdes for accuracy */
+	ret = qcom_xpcs_read(qxpcs, DW_VR_MII_PCS_DIG_CTRL1);
+	if (ret < 0)
+		goto failure;
+
+	qcom_xpcs_write(qxpcs, DW_VR_MII_PCS_DIG_CTRL1, ret | SOFT_RST);
+
+	ret = qcom_xpcs_poll_reset(qxpcs, DW_VR_MII_PCS_DIG_CTRL1, SW_RST_BIT_STATUS);
+
+	if (ret < 0) {
+		XPCSDBG("Poll reset failed\n");
+		goto recover;
+	}
+
+	if (!qxpcs->intr_en) {
+		switch (qxpcs->phy_interface) {
+		case PHY_INTERFACE_MODE_USXGMII:
+			/* Check for Link status */
+			ret = qcom_xpcs_poll_bit_set(qxpcs,
+						     DW_SR_MII_MMD_STS, DW_SR_MII_STS_LINK_STS);
+			if (ret < 0)
+				goto recover;
+			fallthrough;
+		case PHY_INTERFACE_MODE_10GBASER:
+		case PHY_INTERFACE_MODE_5GBASER:
+			/* Check for remote Link status */
+			ret = qcom_xpcs_poll_bit_set(qxpcs,
+						     DW_SR_MII_PCS_STS1, DW_SR_XS_PCS_STS1);
+			if (ret < 0) {
+				XPCSDBG("Link is down try to recover\n");
+				goto recover;
+			}
+			/* Check for block lock status */
+			ret = qcom_xpcs_poll_bit_set(qxpcs,
+						     DW_SR_MII_PCS_KR_STS2, DW_LAT_BL);
+			if (ret < 0) {
+				XPCSDBG("DW_LAT_BL goto recover\n");
+				goto recover;
+			}
+			/* Check for block error  status */
+			ret = qcom_xpcs_poll_reset(qxpcs, DW_SR_MII_PCS_KR_STS2, DW_ERR_BLK);
+			if (ret < 0) {
+				XPCSDBG("DW_ERR_BLK goto recover\n");
+				goto recover;
+			}
+		}
+		state->link = true;
+		return 0;
+	}
+
+failure:
+	state->link = false;
+	return ret;
+}
+
 static int xpcs_get_state_c37_usxgmii(struct dw_xpcs_qcom *qxpcs,
 				      struct phylink_link_state *state)
 {
@@ -475,10 +552,18 @@ static void qcom_xpcs_get_state(struct phylink_pcs *pcs,
 
 	switch (compat->an_mode) {
 	case DW_AN_C37_USXGMII:
-		ret = xpcs_get_state_c37_usxgmii(qxpcs, state);
+		if (!qxpcs->intr_en)
+			ret = qcom_xpcs_get_link_status(qxpcs, state);
+		else
+			ret = xpcs_get_state_c37_usxgmii(qxpcs, state);
 		if (ret < 0)
 			XPCSERR("Failed to get USXGMII state\n");
 		break;
+	case DW_10GBASER:
+		ret = qcom_xpcs_get_link_status(qxpcs, state);
+		if (ret < 0)
+			XPCSERR("Failed to get BaseR state\n");
+		break;
 	default:
 		return;
 	}
@@ -542,7 +627,7 @@ static int qcom_xpcs_enable(struct phylink_pcs *pcs)
 	struct dw_xpcs_qcom *qxpcs = phylink_pcs_to_xpcs(pcs);
 	int ret = 0;
 
-	if (qxpcs->pcs_fusa_intr)
+	if (qxpcs->pcs_fusa_intr > 0)
 		ret = qcom_xpcs_fusa_intr_enable(qxpcs);
 
 	return ret;
@@ -552,7 +637,7 @@ static void qcom_xpcs_disable(struct phylink_pcs *pcs)
 {
 	struct dw_xpcs_qcom *qxpcs = phylink_pcs_to_xpcs(pcs);
 
-	if (qxpcs->pcs_fusa_intr)
+	if (qxpcs->pcs_fusa_intr > 0)
 		qcom_xpcs_fusa_intr_disable(qxpcs);
 }
 
diff --git a/drivers/net/pcs/pcs-xpcs-qcom.h b/drivers/net/pcs/pcs-xpcs-qcom.h
index c43b5cf84cff..1196808dd899 100644
--- a/drivers/net/pcs/pcs-xpcs-qcom.h
+++ b/drivers/net/pcs/pcs-xpcs-qcom.h
@@ -2,7 +2,7 @@
 /*
  * Synopsys DesignWare XPCS platform device driver
  *
- * Copyright (c) 2024-2025 Qualcomm Innovation Center, Inc. All rights reserved.
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
  */
 
 #include <linux/pcs-xpcs-qcom.h>
@@ -15,6 +15,10 @@
 #define DW_VENDOR			BIT(15)
 
 /* VR_XS_PCS */
+#define DW_LAT_BL			BIT(15)
+#define DW_TF				BIT(11)
+#define DW_RF				BIT(10)
+#define DW_ERR_BLK			GENMASK(7, 0)
 #define DW_USXGMII_RST			BIT(10)
 #define DW_USXGMII_EN			BIT(9)
 #define DW_VR_XS_PCS_DIG_STS		0x2040 /* SWI name: EMAC0_VR_XS_PCS_DIG_STS */
